File: python/n112_PathSum.py
# https://leetcode-cn.com/problems/path-sum/
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if not root:
            return False
        sum -= root.val
        if root.left is None and root.right is None:
            return sum == 0
        return self.hasPathSum(root.left, sum) or self.hasPathSum(root.right, sum)

    def hasPathSum_dfs(self, root: TreeNode, sum: int) -> bool:
        print(self.dfs_search(root))
        if sum in self.dfs_search(root):
            return True
        else: return False
    
    def dfs_search(self, root: TreeNode):
        path_sum = []
        node_stack = []
        p = root       
        sum = 0
        while node_stack != [] or p is not None:           
            while p is not None: 
                sum += p.val
                node_stack.append((p, sum))                              
                p = p.left
            p = node_stack.pop()
            if p[0].left is None and p[0].right is None:
                path_sum.append(sum) 
            sum = p[1]                
            p = p[0].right
        return path_sum
        
        
if __name__ == "__main__":
    pass
            

File: cpp/n112_PathSum.cpp
// https://leetcode-cn.com/problems/path-sum/
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if (root == NULL) return false;
        if (root -> left == NULL && root -> right == NULL) return sum == root -> val;
        return hasPathSum(root -> left, sum - (root -> val)) || hasPathSum(root -> right, sum - (root -> val));
    }
    
    bool dfs_hasPathSum(TreeNode* root, int sum) {
        if (root == NULL) return false;
        int path_sum = 0;
        stack <pair<TreeNode*, int>> node_stack;
        TreeNode* p = root;
        while (!node_stack.empty() || p != NULL){
            while (p != NULL){
                path_sum += p -> val;
                node_stack.push(pair<TreeNode*, int>(p, path_sum));
                p = p -> left;
            }
            p = node_stack.top().first;
            path_sum = node_stack.top().second;
            node_stack.pop();            
            if (p -> left == NULL && p -> right == NULL && path_sum == sum) return true;          
            p = p -> right;
        }
        return false;
    }
};

int main(){
    return 0;
}